### DISCUSSION TITLE
Document and license expiry date on identification for reminder notifications.

### DISCUSSION THREAD

**Discussion by the-kizz** [dani-garcia/vaultwarden#3758] [Ideas]
Incorporating an expiry date feature for identity documents, such as licenses, would be extremely beneficial. This feature could be leveraged by apps for sending push notifications and reminders as these documents near their expiration dates.

In the absence of such a feature being picked up by third party apps, native alerts could use SMTP as an interim solution, enabling the sending of email alerts when document expiry dates approach.

Additionally, the inclusion of a document number/card number field would be greatly helpful. While I know that this can be currently managed through custom fields, a dedicated field would improve usability and consistency.

The use of a custom field for expiry dates would be harder, considering my aspiration to leverage this data for notification systems integration in the future. The logic would be far more complex using a custom field.

**Comment by matlink** [comment]
The thing is we can't do much server-side as we depend on the fields available at the client-side. So unless the bitwarden clients get such a field, I guess it is irrelevant to implement this.
A more meta thinking: is it the job of Vaultwarden/Bitwarden to monitor the expiration date of documents?

**Reply by the-kizz** [reply]
Would my suggestion of using smtp to send email alerts not work?

i guess there’s risk in decrypting data if using a custom field. 

I feel it could be done. But you’re probably right. It likely needs client support first.

my initial thought was:
	1.	Custom Field Creation: Users would need to create a custom field named “Expiry” (or something similar) when adding or editing an item in Bitwarden. The value of this field should be the expiry date of the document or the license.
	2.	Server-Side Script: Develop a script that runs on the server hosting Vaultwarden. This script should do the following:
	•	Authenticate: Use Bitwarden’s API to authenticate and access the vault.
	•	Fetch Items: Fetch all items from the vault.
	•	Parse Items: Parse each item, looking for the custom field “Expiry.”
	•	Compare Dates: If an “Expiry” field is found, extract the date value and compare it with the current date.
	•	Send Alert: If the expiry date is nearing (the alert threshold can be defined, like 7 days), use SMTP to send an email alert to the user.

**Reply by BlackDex** [reply]
We can't decrypt data on the server side, we do not know the decryption password. So that is not an option.
Only way would be if Bitwarden would send the expire data as a non encrypted value, and then we could only send a mail that an item will expire. not with a name or something else.

In the database it is stored like this:
```json
"ExpMonth":"2.4dKDDUKD7e9RvyQ5ZVcENA==|f8/GZV/sjRk0GUKaEX80KQ==|+MnFewlK9Lex44crySawuVTZ2ovejtY1ViGU1Q8yokg=",
"ExpYear":"2.K5LhfwdYGeIf84OPwPxIVA==|cKMpGVWrd7WrIugXkuKrLQ==|dePGuvMpchhA78CMn++K2CG78N7+XWEteJ33q70B1cM=",
```

We can't decrypt that, so no way to do anything with it. The same goes for custom fields.
So, what you are requesting is not possible on the server side.

On the client side however it should be possible, same as like is done with exposed passwords, reused passwords, weak passwords etc...

**Reply by the-kizz** [reply]
Ok. Makes sense. 

Maybe I’ll do a feature request for Bitwarden, then we can implement it here if it’s picked up. 

I see value to it. Even without push notifications.
Even just logging in and seeing what is approaching. 

Could this be done in the webui? 
could even do some cool things with the colours, as it approaches expiry.

**Reply by matlink** [reply]
> Could this be done in the webui?\ncould even do some cool things with the colours, as it approaches expiry.

Yes it could be. However I believe the vaultwarden's team tries to keep the webvault as pure as possible, meaning they patch the webvault in the last resort.
